apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -409,3 +410,183 @@ export const publishResource = async (
+    .withRecursive("resourcePath", (eb) =>
+      eb
+        .selectFrom("Resource as r")
+        .select([
+          "r.id",
+          "r.title",
+          "r.permalink",
+          "r.parentId",
+          "r.permalink as fullPermalink",
+        ])
+        .where("r.parentId", "is", null)
+        .unionAll(
+          eb
+            .selectFrom("Resource as s")
+            .innerJoin("resourcePath as rp", "s.parentId", "rp.id")
+            .select([
+              "s.id",
+              "s.title",
+              "s.permalink",
+              "s.parentId",
+              sql<string>`CONCAT(rp."fullPermalink", '/', s.permalink)`.as(
+                "fullPermalink",
+              ),
+            ]),
+        ),
+    )
+    .selectFrom("resourcePath as rp")
+    .select(["rp.id", "rp.title", "rp.fullPermalink"])
+    .where("rp.id", "=", resourceId)
+    .executeTakeFirst()
+
+  return result
+}
+
+const getResourcesWithLastUpdatedAt = ({ siteId }: { siteId: number }) => {
+  return db
+    .selectFrom("Resource")
+    .select([
+      "Resource.id",
+      "Resource.title",
+      "Resource.type",
+      "Resource.parentId",
+      // To handle cases where either the resource or the blob is updated
+      sql`GREATEST("Resource"."updatedAt", "Blob"."updatedAt")`.as(
+        "lastUpdatedAt",
+      ),
+    ])
+    .leftJoin("Blob", "Resource.draftBlobId", "Blob.id")
+    .where("Resource.siteId", "=", siteId)
+}
+
+const getResourcesWithFullPermalink = async ({
+  resources,
+}: {
+  resources: Omit<SearchResultResource, "fullPermalink">[]
+}): Promise<SearchResultResource[]> => {
+  return await Promise.all(
+    resources.map(async (resource) => ({
+      ...resource,
+      fullPermalink: await getWithFullPermalink({
+        resourceId: resource.id,
+      }).then((r) => r?.fullPermalink ?? ""),
+    })),
+  )
+}
+
+export const getSearchResults = async ({
+  siteId,
+  query,
+  offset,
+  limit,
+}: {
+  siteId: number
+  query: string
+  offset: number
+  limit: number
+}): Promise<{
+  totalCount: number | null
+  resources: SearchResultResource[]
+}> => {
+  const searchTerms: string[] = Array.from(
+    new Set(query.trim().toLowerCase().split(/\s+/)),
+  )
+
+  const queriedResources = getResourcesWithLastUpdatedAt({
+    siteId: Number(siteId),
+  })
+    .where("Resource.type", "in", [
+      // only show user-viewable resources (excluding root page, folder meta etc.)
+      ResourceType.Page,
+      ResourceType.Folder,
+      ResourceType.Collection,
+      ResourceType.CollectionLink,
+      ResourceType.CollectionPage,
+    ])
+    .where((eb) =>
+      eb.or([
+        ...searchTerms.map((searchTerm) =>
+          // Match if the search term is at the start of the title
+          eb("Resource.title", "ilike", `${searchTerm}%`).or(
+            // Match if the search term is in the middle of the title (after a space)
+            eb("Resource.title", "ilike", `% ${searchTerm}%`),
+          ),
+        ),
+      ]),
```

Comment:
minor nit: I think we don't need to spread the array? (editing on GitHub so the indentation might be incorrect)

```suggestion
      eb.or(searchTerms.map((searchTerm) =>
          // Match if the search term is at the start of the title
          eb("Resource.title", "ilike", `${searchTerm}%`).or(
            // Match if the search term is in the middle of the title (after a space)
            eb("Resource.title", "ilike", `% ${searchTerm}%`),
          ),
        )),
```